govcd/org_oidc.go

```diff
@@ -0,0 +1,314 @@
+/*
+ * Copyright 2024 VMware, Inc.  All rights reserved.  Licensed under the Apache v2 License.
+ */
+
+package govcd
+
+import (
+	"bytes"
+	"encoding/xml"
+	"fmt"
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+)
+
+// GetOpenIdConnectSettings retrieves the current OpenID Connect settings for a given Organization
+func (adminOrg *AdminOrg) GetOpenIdConnectSettings() (*types.OrgOAuthSettings, error) {
+	result, err := oidcExecuteRequest(adminOrg, http.MethodGet, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+// SetOpenIdConnectSettings sets the OpenID Connect configuration for a given Organization. If the well-known configuration
+// endpoint is provided, the configuration is automatically retrieved from that URL.
+// If other fields have been set in the input structure, the corresponding values retrieved from the well-known endpoint are overridden.
+// If there are no fields informed, the configuration retrieved from the well-known configuration endpoint is applied as-is.
+// ClientId and ClientSecret properties are always mandatory, with and without well-known endpoint.
+// This method returns an error if the settings can't be saved in VCD for any reason or if the provided settings are wrong.
+func (adminOrg *AdminOrg) SetOpenIdConnectSettings(settings types.OrgOAuthSettings) (*types.OrgOAuthSettings, error) {
+	if settings.ClientId == "" {
+		return nil, fmt.Errorf("the Client ID is mandatory to configure OpenID Connect")
+	}
+	if settings.ClientSecret == "" {
+		return nil, fmt.Errorf("the Client Secret is mandatory to configure OpenID Connect")
+	}
+	if settings.WellKnownEndpoint != "" {
+		err := oidcValidateConnection(adminOrg.client, settings.WellKnownEndpoint)
+		if err != nil {
+			return nil, err
+		}
+		wellKnownSettings, err := oidcConfigureWithEndpoint(adminOrg.client, adminOrg.AdminOrg.HREF, settings.WellKnownEndpoint)
+		if err != nil {
+			return nil, err
+		}
+		// The following conditionals allow users to override the well-known automatic configuration values with their own,
+		// mimicking what users can do in UI.
+		// If an attribute was not set in the input settings, we pick the value that the well-known endpoint gave for that attribute,
+		// but if it was explicitly set by the user, we take that one instead (overriding the well-known one).
+		if settings.AccessTokenEndpoint == "" {
```

Comment:
Can we use [`cmp.Or`](https://pkg.go.dev/cmp#Or) or a similar implementation for all these settings?

Reply:
That's a really good idea, I was not aware of that new function.

Heads-up for reviewers that already approved: @Didainius @lvirbalas, I have bumped Go to 1.22:
https://github.com/vmware/go-vcloud-director/pull/671/commits/63e84662107d1d29dc5f6d1e3359d981e9229c3c

Reply:
I'm checking this: https://github.com/vmware/go-vcloud-director/actions/runs/9077347647/job/24942032705?pr=671
It seems I broke `make static`, although it works for me locally.

Reply:
Fixed by bumping staticcheck to 2023.1.7 (the changelog stated this problem: _"The released binaries for this version have been built with Go 1.22 and should no longer panic when checking code targeting Go 1.22"_)
https://github.com/vmware/go-vcloud-director/pull/671/commits/7cbe1997140cf66d36509366e3648ea2484ef706